vdev_disk: try harder to ensure IO alignment rules

robn opened the pull request:
_[Sponsors: Klara, Inc., Wasabi Technology, Inc.]_

### Motivation and Context

Ongoing IO alignment issues on Linux with dm-crypt.

Closes: #16631, #15646, #15533, #14533.
Maybe also: #10094

### Description

The key insight from out of #16631 is that blocks data must not cross physical page boundaries. So, this updates the alignment check function, then changes the fallback allocation method to ensure page alignment.

> I'm not very happy with `abd_alloc_scatter()`. It feels like a big call to just create it and assert that its always page-aligned. But it's probaby right, and the ABD interface as a whole is kinda of a mess these days, and there really isn't an easier allocator to reach for that will do page alignment (even using `alloc_pages()` needs a bunch of ceremony that `abd_alloc_chunks()` already does. So it doesn't seem _too_ bad.

Update: @amotin educated me on why the existing `abd_alloc_for_io()` was actually what I wanted. So this has been changed for that, and `abd_alloc_scatter()` no longer included.

> I did notice through this that we don't honour the "minimum IO" device limit at all. Most things don't care, but as usual, dm-crypt does, because the block size is part of the crypto. So creating an `ashift=9` pool over a 4K dm-crypt device just throws IO errors. I think it always did though (I can't see how it couldn't have), so I'm not going to worry too much for now. It does show that we need to work to minimum limits, which will mean eg not doing <4K accesses, but that seems like it's gonna be work all over the stack (labels, gang headers, etc)... some other time!

Update: Turns out this was a long-standing bug with ZFS, in that it should never be possible to use an ashift smaller than LBS. So now we do check for this case, and #16690 will make it impossible to get into that situation.

### How Has This Been Tested?

The test case in #16631 doesn't trip the bug anymore. If I set `zfs_vdev_disk_classic=1` it blows up pretty quick. My test cases from #15588 (forcing ganging) continue to work correctly.

> Full ZTS run in progress, but it's all looking pretty good so far.

Update: ZTS ran to completion.

Update: I've now run the #16631 test case with various combinations of `--sector-size=` to `cryptsetup luksCreate` and `-o ashift=` to `zpool create` (always with ashift larger than sector size). Nothing tripped.

### Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -834,38 +823,67 @@ vbio_completion(struct bio *bio)
 	/*
-	 * Note if we're taking less than a full block, so we can check it
-	 * above on the next call.
+	 * TODO: Ideally we would require that len is a multiple of LBS.
+	 *       To do that we need to not allow setting ashift= smaller
+	 *       than the LBS. Often, it will end up failing this check
```
ashift smaller than LBS should never happen, since attempt of ZFS write that would either fail or cause read-modify-write, if something below is expected to handle it, but we don't want it for several reasons.  If there is a way for ZFS to misread LUKS LBS, then it much be fixed, and user should not be allowed to set ashift smaller than LBS.  That is how it is on FreeBSD and how I believe it must be done.

robn replied:
I observed that creating a 4K dm-crypt map, then doing `zpool create -o ashift=9 tank /dev/mapper/mycrypt` spewed zio errors. Looking at the `vdev_t` for the device, it had `vdev_logical_shift=12`, `vdev_physical_shift=12`, but `vdev_ashift=9`. As far as I can tell, `vdev_open` is responsible for setting these based on what it gets from the vdev driver. So, ZFS is not "misreading LUKS LBS", but something weird is happening further upstack, nothing to do with this.

If that was fixed, then I would do what I originally had in place of this TODO comment: making sure the data length within this page was an even multiple of the LBS.

behlendorf replied:
When the user provides an ashift it looks like we're only enforcing that it be `>= ASHIFT_MIN` and `<= ASHIFT_MAX`.  We should additionally check that it's `>= vdev_logical_shift`  and fail the `zpool create` if it's not.  This should do the trick, but I haven't had a chance to test it.

```diff
diff --git a/module/zfs/vdev.c b/module/zfs/vdev.c
index bcab46c63b..60a2e3e409 100644
--- a/module/zfs/vdev.c
+++ b/module/zfs/vdev.c
@@ -2222,6 +2222,7 @@ vdev_open(vdev_t *vd)
                        vd->vdev_attaching = B_FALSE;
                }
                if (vd->vdev_ashift != 0 && (vd->vdev_ashift < ASHIFT_MIN ||
+                   vd->vdev_ashift < vd->vdev_logical_ashift ||
                    vd->vdev_ashift > ASHIFT_MAX)) {
                        vdev_set_state(vd, B_TRUE, VDEV_STATE_CANT_OPEN,
                            VDEV_AUX_BAD_ASHIFT);
```

robn replied:
I wonder if this is going to break for anyone? At least, I know that in the block layer block sizes are all converted to counts of 512b sectors, and it's up to drivers to enforce. I haven't checked specifically but I wouldn't be at all surprised to find that a great many drivers or drives don't actually care or know how to handle it.

I still think we should do it, but it feels like the kind of thing that needs a bunch of thinking first?

amotin replied:
I am really surprised that it does not work as I expected.  I don't think ashift property set on a pool should result in any error, rather it should be ignored at the time of vdev creation for vdevs where it makes no sense.  I'll take a look on it tomorrow, since today I am too tired and jet-lagged.  If though pool was created somehow created with ashift smaller than LBS (likely somebody messed up with VM configuration or some very weird disk driver) -- that must be fatal at the time of pool import.

amotin replied:
I think this should fix the issue: https://github.com/openzfs/zfs/pull/16690 .

robn replied:
Nice. Meanwhile, I've removed the TODO comment and reintroduced the multiple-of-LBS check. See below.

tonyhutter reviewed (approved):
A lot of this is over my head, but I don't see any surface level issues.  I also tested it with e5d1f6816758f3ab1939608a9a45c18ec36bef41 and it passed.   It may need to be re-based though.

amotin commented on module/os/linux/zfs/vdev_disk.c:
```diff
@@ -916,37 +933,35 @@ vdev_disk_io_rw(zio_t *zio)
+		 * alignment.
 		 */
-		abd = abd_get_from_buf(buf, zio->io_size);
+		abd = abd_alloc_scatter(zio->io_size,
+		    zio->io_abd->abd_flags & ABD_FLAG_META);
```
I don't think we really need need abd_alloc_scatter() here or anywhere at all, unless you strictly intend to allocate non-KVA-mapped pages, which abd_alloc() allocates by default for big buffers and which make no sense for small ones.  We should never receive here buffers with size not multiple of ashift, and for those abd_alloc() should always do the right thing.  Or we actually have abd_alloc_for_io(), which should do the best thing for each platform exactly for this use case.

robn replied:
I need to guarantee an allocation aligned to the start of a page (KVA-mapped is irrelevant). We don't have a convenient way to get that guarantee in any of our standard allocators.

The problem is that `abd_alloc()` will not necessarily always allocate scatter; it depends on the tuneables that drive `abd_size_alloc_linear()`. For linear, we know that `zio_buf_alloc()` and `zio_data_buf_alloc()` won't necessarily return page-aligned allocations.

If I'm wrong, and we do have an allocator (or set of flags) that is _guaranteed_ to return page-aligned memory, point me at it. I couldn't see it. Otherwise, I have to make one (or adjust an existing one). Options seemed to be:
- round up the allocation size to something we know would get a page-aligned allocation from an existing allocator
- add an explicit "aligned" allocator (or flag to an existing allocator), something like `posix_memalign`
- use `__vmalloc` (that is, `spl_vmalloc()) which appears to guarantee page alignment, but is also fully mapped
- use `alloc_pages()`, which isn't mapped, and is a bit more work

I ended up deciding that `alloc_pages()` is nicer than `__vmalloc()` because apart from the brief moment while copying data from/to the ZIO ABD, those pages don't need to be mapped, and never need to be contiguous. So it should be much kinder for memory pressure. But, it still needs to be wrapped in an ABD, which is more fiddly.

And then I saw `abd_alloc_chunks()` was doing _exactly_ what I needed to do anyway, so I took the shorter route and just exposed its internals.

Though, looking again, I see that we now have `abd_alloc_from_pages()` to wrap an array of page pointers in an ABD, so I guess I could just do it that way (it's very new, so not yet internalised into my standard toolkit heh). I'll give that a go shortly.

I do think that probably `abd_alloc_for_io()` should be trying harder to sort out the proper alignment needed, or pick something that works for everything. Maybe that's not so easy to do if you don't know which device the IO is going to be sent to. I don't know yet. I do think the ABD API is kind of a nightmare; it acts like its abstracting over the details of how a buffer lives in memory, but then in lots of places we have to ask "are you linear?" or "are you scatter?" or whatever. I do intend to try and fix a lot of that, but I don't want to do it here. I freely admit that `abd_alloc_scatter()` is making that mess worse, not better.

Ultimately here, I just need aligned memory. I don't much care where it comes from. I'll try `alloc_pages()` + `abd_alloc_from_pages()` soon. Any other options welcome.

amotin replied:
You don't need page alignment.  You need a (potentially) virtually contiguous buffer, aligned to ashift (or PAGESIZE if ashift is bigger than it).  And since my last refactor at https://github.com/openzfs/zfs/commit/bd7a02c251d8c119937e847d5161b512913667e6 abd_alloc() (and on Linux abd_alloc_for_io() also as its wrapper) do exactly that, if you provide them with size multiple of ashift, which you will.  The problem workarounded by https://github.com/openzfs/zfs/pull/16678 is that buffer initially allocated as 6KB then truncated to 4KB, that breaks this alignment convention.  If you allocate buffer directly as needed with ashift-multiple size, it will be properly aligned.  Trust me.

robn replied:
I suppose no, I don't technically need page-aligned, but that was the only cheap and easy rule to deal with what I was seeing.

So, to be clear: are you saying if I ask for io_size from abd_alloc_for_io, I will never get a LBS block crossing a memory page? If so, very excellent, and I'm gonna add commentary and assertions to that effect so I never have to experience this ever again 😆

amotin replied:
Yes, that is what I am saying. Obviously while LBS <= PAGESIZE.  I remember I already wondered once what will Linux do when disks with LBS > PAGESIZE appear one day.

robn replied:
I'm sure "cry".

Last push removed `abd_alloc_scatter()` and switched to `abd_alloc_for_io()`. Seems ok!

amotin reviewed (approved):
Looks good to me except the TODO comment.

robn commented:
Last push removed the TODO comment, added a check to ensure that the data len within the page is a multiple of LBS, and adjusted the `page_alignment` test to match.

allanjude reviewed (commented):
Reviewed-by: Allan Jude <<allan@klarasystems.com>>

wiesl commented:
@robn @behlendorf @allanjude @amotin @tonyhutter

First of all, thank you that this bug has been tracked down and finally fixed.

What I'm asking me:
1. What were the consequences of that unaligned writes?
2. Can previous unaligned writes (which are hit by the bug) made before this bugfix still have negative consequences in the future (e.g. file system corruption)?
3. If scrub is now ok: Is it 100% proven, that no negative consequences can occour due this bug in the future? Or is there still an (unlikely) chance that prior unaligned writes before the bugfix might have negative consequences?
4. If there are still negative consequences possible how can we detect them and fix them?
(e.g. has scrub to be enhanced?)

So overall: might a zpool be in a state where we don't want it in that state for future file system operations.

Just want to make sure that data reliability of ZFS is high :-)

Thnx.
